packages/synapse-react-client/src/components/QueryContext/QueryContext.tsx

```diff
@@ -23,11 +24,33 @@ export type LockedColumn = {
   value?: string
 }
 
-export type QueryContextType = {
+type OptionalQueryBundleRequestFields = keyof Pick<
+  QueryResultBundle,
+  | 'queryResult'
+  | 'queryCount'
+  | 'selectColumns'
+  | 'maxRowsPerPage'
+  | 'columnModels'
+  | 'facets'
+  | 'sumFileSizes'
+  | 'combinedSql'
+>
+
+export type QueryContextType<
+  TIncludedFields extends OptionalQueryBundleRequestFields = never,
+  ExcludeOtherFields extends true | false = false,
+> = {
   /** The entity being queried. Will be undefined while initially fetching */
   entity: Table | undefined
   /** The query results, which will be undefined while initially fetching a new bundle, but will not be unloaded when fetching new pages */
-  data: QueryResultBundle | undefined
+  data:
+    | Omit<
+        SetRequired<QueryResultBundle, TIncludedFields>,
+        ExcludeOtherFields extends true
+          ? Exclude<OptionalQueryBundleRequestFields, TIncludedFields>
+          : never
+      >
```

Comment:
`SetRequired<QueryResultBundle, TIncludedFields>` makes the specified fields non-optional (e.g. we can enforce that the columnModels were returned).

The `Omit` removes the other fields if `ExcludeOtherFields` is true. If it's false, the other fields will remain optional aka possibly undefined.